### ISSUE TITLE
write collect Tracing events to disk after each iteration

### ISSUE THREAD

**Issue by worenga** [sitespeedio/browsertime#308]
When performing a lot of repetitions (in my case -n 31) for some pages which have a very large amount of resources (>256) i noticed nodejs crashes frequently due to out of memory errors (my instances have around 4GB of main memory). When I looked at memory consumption i saw that it was increasing with the amount of runs.

I've disabled  `--chrome.collectPerfLog` and `--chrome.collectTracingEvents` and the issue no longer appears.

I guess it is because the results are kept in memory until all runs are finished, a viable workaround would be to dump the results after each iteration (like the netlog)

```
/bin/browsertime.js  --chrome.binaryPath /home/bewo/chrome_sources/Default/chrome -n 31 --chrome.args="--disable-quic" --chrome.args="--ssl-version-max=tls1.2" --viewPort"1200x960" --chrome.collectTracingEvents  --chrome.collectPerfLog  --chrome.collectNetLog  --resultDir results_09042017/latency_25_fraction_256_push//push https://h2pushsrv/pushshare/256/?push
Timeout is 3720
[2017-04-09 20:30:49] Running chrome for url: https://h2pushsrv/pushshare/256/?push
[2017-04-09 20:30:49] Testing url https://h2pushsrv/pushshare/256/?push run 1
[2017-04-09 20:30:57] Testing url https://h2pushsrv/pushshare/256/?push run 2
[2017-04-09 20:31:05] Testing url https://h2pushsrv/pushshare/256/?push run 3
[2017-04-09 20:31:13] Testing url https://h2pushsrv/pushshare/256/?push run 4
[2017-04-09 20:31:22] Testing url https://h2pushsrv/pushshare/256/?push run 5
[2017-04-09 20:31:30] Testing url https://h2pushsrv/pushshare/256/?push run 6
[2017-04-09 20:31:38] Testing url https://h2pushsrv/pushshare/256/?push run 7
[2017-04-09 20:31:46] Testing url https://h2pushsrv/pushshare/256/?push run 8
[2017-04-09 20:31:53] Testing url https://h2pushsrv/pushshare/256/?push run 9
[2017-04-09 20:32:01] Testing url https://h2pushsrv/pushshare/256/?push run 10
[2017-04-09 20:32:09] Testing url https://h2pushsrv/pushshare/256/?push run 11
[2017-04-09 20:32:17] Testing url https://h2pushsrv/pushshare/256/?push run 12
[2017-04-09 20:32:25] Testing url https://h2pushsrv/pushshare/256/?push run 13
[2017-04-09 20:32:33] Testing url https://h2pushsrv/pushshare/256/?push run 14
[2017-04-09 20:32:41] Testing url https://h2pushsrv/pushshare/256/?push run 15
[2017-04-09 20:32:49] Testing url https://h2pushsrv/pushshare/256/?push run 16
[2017-04-09 20:32:58] Testing url https://h2pushsrv/pushshare/256/?push run 17
[2017-04-09 20:33:06] Testing url https://h2pushsrv/pushshare/256/?push run 18
[2017-04-09 20:33:14] Testing url https://h2pushsrv/pushshare/256/?push run 19
[2017-04-09 20:33:22] Testing url https://h2pushsrv/pushshare/256/?push run 20
[2017-04-09 20:33:31] Testing url https://h2pushsrv/pushshare/256/?push run 21
[2017-04-09 20:33:39] Testing url https://h2pushsrv/pushshare/256/?push run 22
[2017-04-09 20:33:47] Testing url https://h2pushsrv/pushshare/256/?push run 23
[2017-04-09 20:33:56] Testing url https://h2pushsrv/pushshare/256/?push run 24
[2017-04-09 20:34:05] Testing url https://h2pushsrv/pushshare/256/?push run 25
[2017-04-09 20:34:15] Testing url https://h2pushsrv/pushshare/256/?push run 26
[2017-04-09 20:34:26] Testing url https://h2pushsrv/pushshare/256/?push run 27
[2017-04-09 20:34:47] Testing url https://h2pushsrv/pushshare/256/?push run 28
[2017-04-09 20:35:17] Testing url https://h2pushsrv/pushshare/256/?push run 29
[2017-04-09 20:36:03] Testing url https://h2pushsrv/pushshare/256/?push run 30
[2017-04-09 20:37:04] Testing url https://h2pushsrv/pushshare/256/?push run 31
[2017-04-09 20:38:12] 2 requests, 45.23 kb, BackeEndTime: 247ms (±1.13ms), firstPaint: 1.20s (±10.23ms), DOMContentLoaded: 1.20s (±10.20ms), Load: 1.20s (±10.23ms), rumSpeedIndex: 1203 (±10.23) (31 runs)


<--- Last few GCs --->

[13813:0x3caf890]   603173 ms: Mark-sweep 2047.0 (2472.5) -> 2045.5 (2475.5) MB, 1174.6 / 0.0 ms  allocation failure GC in old space requested
[13813:0x3caf890]   604331 ms: Mark-sweep 2045.5 (2475.5) -> 2045.4 (2421.0) MB, 1157.4 / 0.0 ms  last resort 
[13813:0x3caf890]   605380 ms: Mark-sweep 2045.4 (2421.0) -> 2045.4 (2398.5) MB, 1048.2 / 0.0 ms  last resort 


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x11fa020266a1 <JS Object>
    1: fromString(aka fromString) [buffer.js:207] [pc=0x445fc0cbcfa](this=0x305fa4d02311 <undefined>,string=0x1affa2df9ec1 <Very long string[1752702]>,encoding=0x11fa0202f2e1 <String[4]: utf8>)
    2: from [buffer.js:104] [pc=0x445fc0cb91c](this=0x23318cb3cc9 <JS Function Buffer (SharedFunctionInfo 0x11fa0204b469)>,value=0x1affa2df9ec1 <Very long string[1752702]>,encodingOrOffset=0x11fa0202f2e1...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [node]
 2: 0x134e6cc [node]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node]
 5: v8::internal::Factory::NewRawOneByteString(int, v8::internal::PretenureFlag) [node]
 6: v8::internal::String::SlowFlatten(v8::internal::Handle<v8::internal::ConsString>, v8::internal::PretenureFlag) [node]
 7: v8::internal::String::Flatten(v8::internal::Handle<v8::internal::String>, v8::internal::PretenureFlag) [node]
 8: v8::String::WriteUtf8(char*, int, int*, int) const [node]
 9: node::StringBytes::Write(v8::Isolate*, char*, unsigned long, v8::Local<v8::Value>, node::encoding, int*) [node]
10: node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::encoding) [node]
11: node::Buffer::CreateFromString(v8::FunctionCallbackInfo<v8::Value> const&) [node]
12: v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) [node]
13: 0xabfbed [node]
14: v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) [node]
15: 0x445faf843a7
```

**Comment by soulgalore** [comment]
Hi @worenga something like that would be the long time solution, for now increasing memory like we do inhttps://github.com/sitespeedio/sitespeed.io/blob/master/docker/scripts/start.sh#L25 could help (but you maybe done that already?). The default nodejs memory limit is something like 2g so increasing that could help, but I guess you will need a larger instance.

Best
Peter

**Comment by tobli** [comment]
It's a use case that stretches the current design of the Browsertime api. Currently everything is collected in memory. The Browsertime cli uses the same api (that returns json data and screenshots) as does external clients such as sitespeed.io. For really large datasets a different approach (perhaps callback or event based) might be a better fit. My suggestion for next step would be to start collecting use cases to consider for a next-gen design. Maybe create a meta-issue for that?

**Comment by tobli** [comment]
That said, I'm also open to short term options for controlling how/when data is stored

**Comment by soulgalore** [comment]
Actually, this is fixed in master and will come in 3.0 later this year!

**Comment by soulgalore** [comment]
fixed in master
